### PR TITLE
RavenDB-16958 - Handle external / cluster tcp connection between servers properly

### PR DESCRIPTION
**Pull request by lastav5**
https://issues.hibernatingrhinos.com/issue/RavenDB-16958

### CODE REVIEW

File: src/Raven.Client/Documents/Subscriptions/SubscriptionWorker.cs [hunk 1]
```diff
@@ -278,6 +259,22 @@ await requestExecutor.ExecuteAsync(_redirectNode, null, context, command, should
...
+                Database = _store.GetDatabase(_dbName),
+                Operation = TcpConnectionHeaderMessage.OperationTypes.Subscription,
+                Version = SubscriptionTcpVersion ?? TcpConnectionHeaderMessage.SubscriptionTcpVersion,
+                ReadResponseAndGetVersionCallbackAsync = ReadServerResponseAndGetVersionAsync,
+                DestinationNodeTag = CurrentNodeTag,
+                DestinationUrl = chosenUrl,
+                DestinationServerGuid = tcpInfo.ServerGuid
+            };
+
+            return await TcpNegotiation.NegotiateProtocolVersionAsync(context, stream, parameters).ConfigureAwait(false);
```
Comment (ppekrol) [inline src/Raven.Client/Documents/Subscriptions/SubscriptionWorker.cs:279]: you can remove `async` and just return a task. There are no other actions in this method.

File: src/Raven.Client/Util/TcpUtils.cs [hunk 2]
```diff
@@ -1,14 +1,16 @@
...
 
 #if !(NETSTANDARD2_0 || NETSTANDARD2_1 || NETCOREAPP2_1)
 #define SSL_STREAM_CIPHERSUITESPOLICY_SUPPORT
 #endif
 
-using System;
-using System.Diagnostics;
+ using System;
+using System.Collections.Generic;
+ using System.Diagnostics;
```
Comment (ppekrol) [inline src/Raven.Client/Util/TcpUtils.cs]: sth happened here to formatting

File: src/Raven.Server/ServerWide/ServerStore.cs [hunk 3]
```diff
@@ -2166,6 +2166,8 @@ public Guid GetServerId()
             return _env.DbId;
         }
 
+        public Guid ServerGuid => GetServerId();
```
Comment (ppekrol) [inline src/Raven.Server/ServerWide/ServerStore.cs]: ServerId

File: src/Raven.Client/ServerWide/Tcp/TcpConnectionHeader.cs [hunk 4]
```diff
@@ -47,6 +47,8 @@ public DynamicJsonValue ToJson()
 
         public string SourceNodeTag { get; set; }
 
+        public string ServerGuid; //TODO: change to dbguid?
```
Comment (karmeli87) [inline src/Raven.Client/ServerWide/Tcp/TcpConnectionHeader.cs]: the comment is relevant?

File: src/Raven.Client/Util/TcpUtils.cs [hunk 5]
```diff
@@ -1,14 +1,16 @@
...
 #define TCP_CLIENT_CANCELLATIONTOKEN_SUPPORT
 #endif
 
 #if !(NETSTANDARD2_0 || NETSTANDARD2_1 || NETCOREAPP2_1)
 #define SSL_STREAM_CIPHERSUITESPOLICY_SUPPORT
 #endif
 
-using System;
-using System.Diagnostics;
+ using System;
```
Comment (karmeli87) [inline src/Raven.Client/Util/TcpUtils.cs]: formatting

File: src/Raven.Client/Util/TcpUtils.cs (hunk 2 above)
Comment (karmeli87) [inline src/Raven.Client/Util/TcpUtils.cs]: formatting

File: src/Raven.Client/Util/TcpUtils.cs [hunk 6]
```diff
@@ -1,14 +1,16 @@
-﻿#if !(NETSTANDARD2_0 || NETSTANDARD2_1 || NETCOREAPP2_1 || NETCOREAPP3_1)
+#if !(NETSTANDARD2_0 || NETSTANDARD2_1 || NETCOREAPP2_1 || NETCOREAPP3_1)
```
Comment (karmeli87) [inline src/Raven.Client/Util/TcpUtils.cs:1]: what changed here?

File: src/Raven.Client/Util/TcpUtils.cs [hunk 7]
```diff
@@ -36,10 +40,14 @@ private static void SetTimeouts(TcpClient client, TimeSpan timeout)
             client.SendTimeout = (int)timeout.TotalMilliseconds;
             client.ReceiveTimeout = (int)timeout.TotalMilliseconds;
         }
-
-        internal static async Task<(TcpClient Client, string Url)> ConnectSocketAsync(
-            TcpConnectionInfo connection,
-            TimeSpan timeout,
+        internal static async Task<(TcpClient, Stream, string, TcpConnectionHeaderMessage.SupportedFeatures)> ConnectAndWrapWithSslAsReplication(
```
Comment (karmeli87) [inline src/Raven.Client/Util/TcpUtils.cs]: Please name the return params something like:
`Task<(TcpClient Client, Stream Stream, string Url, TcpConnectionHeaderMessage.SupportedFeatures SupportedFeatures)>`

File: src/Raven.Client/Util/TcpUtils.cs [hunk 8]
```diff
@@ -241,51 +256,91 @@ private static TcpClient NewTcpClient(TimeSpan? timeout, bool useIPv6)
...
-#if TCP_CLIENT_CANCELLATIONTOKEN_SUPPORT
-                ,
-                token: token
-#endif
-                )
-                .ConfigureAwait(false);
-
-            return (tcpClient, info.Url);
+            //Should not reach here
+            throw new Exception();
```
Comment (karmeli87) [inline src/Raven.Client/Util/TcpUtils.cs]: Let's put a message here + `Debug.Assert(false," ... ")` before the throw

File: src/Raven.Client/Util/TcpUtils.cs [hunk 9]
```diff
@@ -241,51 +256,91 @@ private static TcpClient NewTcpClient(TimeSpan? timeout, bool useIPv6)
...
 #endif
             )
         {
-            TcpClient tcpClient;
+            TcpClient tcpClient = null;
+            Stream stream = null;
+            TcpConnectionHeaderMessage.SupportedFeatures supportedFeatures = null;
 
-            if (info.Urls != null)
+            string[]  infoUrls = info.Urls == null ? new string[] {} : info.Urls.Append(info.Url).ToArray();
```
Comment (karmeli87) [inline src/Raven.Client/Util/TcpUtils.cs]: if `info.Urls` is `null`, what will happened here? no connection will be tried? not even the `info.Url`?

File: src/Raven.Server/Documents/Handlers/Debugging/NodeDebugHandler.cs [hunk 10]
```diff
@@ -133,41 +138,83 @@ private async Task<PingResult> PingOnce(string url)
...
+                
+                using (ServerStore.ContextPool.AllocateOperationContext(out JsonOperationContext context))
                 {
-                    var info = await ReplicationUtils.GetTcpInfoAsync(url, null, "PingTest", ServerStore.Engine.ClusterCertificate, cts.Token);
-                    result.TcpInfoTime = sp.ElapsedMilliseconds;
-                    using (var tcpClient = await TcpUtils.ConnectAsync(info.Url, ServerStore.Engine.TcpConnectionTimeout, token: cts.Token).ConfigureAwait(false))
-                    await using (var stream = await TcpUtils.WrapStreamWithSslAsync(tcpClient, info, ServerStore.Engine.ClusterCertificate, Server.CipherSuitesPolicy, ServerStore.Engine.TcpConnectionTimeout, cts.Token).ConfigureAwait(false))
-                    using (ServerStore.ContextPool.AllocateOperationContext(out JsonOperationContext context))
+
+                    List<string> log = new();
```
Comment (karmeli87) [inline src/Raven.Server/Documents/Handlers/Debugging/NodeDebugHandler.cs:156]: Is that needed? you already allocate the `Log` in the `PingResult`.

File: src/Raven.Server/Documents/Handlers/Debugging/NodeDebugHandler.cs [hunk 11]
```diff
@@ -133,41 +138,83 @@ private async Task<PingResult> PingOnce(string url)
...
+                                        logs?.Add(result.Error);
+                                        throw new AuthorizationException(result.Error);
+                                }
+                            }
+                        }
+                        catch (AuthorizationException e)
                         {
-                            result.SendTime = sp.ElapsedMilliseconds;
-                            context.Write(writer, msgJson);
+                            throw e;
```
Comment (karmeli87) [inline src/Raven.Server/Documents/Handlers/Debugging/NodeDebugHandler.cs]: do simply `throw`

Comment (karmeli87) [reply src/Raven.Server/Documents/Handlers/Debugging/NodeDebugHandler.cs]: and why does it need a special treatment?

File: src/Raven.Server/Documents/Handlers/Debugging/NodeDebugHandler.cs [hunk 12]
```diff
@@ -133,41 +138,83 @@ private async Task<PingResult> PingOnce(string url)
...
                         }
-                        using (var response = await context.ReadForMemoryAsync(stream, "cluster-ConnectToPeer-header-response"))
+                        catch (Exception e)
                         {
-                            JsonDeserializationServer.TcpConnectionHeaderResponse(response);
-                            result.ReceiveTime = sp.ElapsedMilliseconds;
+                            result.Error = e.ToString();
+                            logs?.Add($"Error occurred while attempting to negotiate with the server. {e.Message}");
+                            throw new Exception("Error occurred while attempting to negotiate with the server.", e);
```
Comment (karmeli87) [inline src/Raven.Server/Documents/Handlers/Debugging/NodeDebugHandler.cs]: let's re-throw it as is and log additional info.

File: src/Raven.Server/RavenServer.cs [hunk 13]
```diff
@@ -298,6 +296,8 @@ void ConfigureKestrel(KestrelServerOptions options)
                     throw;
                 }
 
+                _tcpListenerStatus = StartTcpListener();
```
Comment (karmeli87) [inline src/Raven.Server/RavenServer.cs]: why this change?

File: src/Raven.Server/Web/System/TestConnectionHandler.cs [hunk 14]
```diff
@@ -41,51 +44,65 @@ await using (var writer = new AsyncBlittableJsonTextWriter(context, ResponseBody
...
         }
 
-        private static async ValueTask WriteOperationHeaderToRemote(AsyncBlittableJsonTextWriter writer, TcpConnectionHeaderMessage.OperationTypes operation, string databaseName)
+        private static async ValueTask WriteOperationHeaderToRemote(AsyncBlittableJsonTextWriter writer, TcpConnectionHeaderMessage.OperationTypes operation,
+            string databaseName, string destinationServerGuid)
         {
-           writer.WriteStartObject();
-           {
+            writer.WriteStartObject();
+            {
```
Comment (karmeli87) [inline src/Raven.Server/Web/System/TestConnectionHandler.cs]: the `{` not needed?

File: src/Raven.Client/Util/TcpUtils.cs (hunk 9 above)
Comment (ppekrol) [inline src/Raven.Client/Util/TcpUtils.cs]: maybe better to use a hashset here with ordinal ignore case comparer? Do we care about the order?

File: src/Raven.Client/Util/TcpUtils.cs (hunk 8 above)
Comment (ppekrol) [inline src/Raven.Client/Util/TcpUtils.cs]: Write that also inside the exception message: `Should not happen!`

File: src/Raven.Server/Documents/Handlers/Debugging/NodeDebugHandler.cs (hunk 11 above)
Comment (ppekrol) [inline src/Raven.Server/Documents/Handlers/Debugging/NodeDebugHandler.cs]: just `throw`
rethrowing is changing the stack

File: src/Raven.Server/RavenServer.cs [hunk 15]
```diff
@@ -1795,6 +1795,10 @@ private void StartListeners(string host, int port, TcpListenerStatus status)
                         errors.Add(new IOException(msg, ex));
                         if (Logger.IsOperationsEnabled)
                             Logger.Operations(msg, ex);
+                        
+                        ServerStore.NotificationCenter.Add(AlertRaised.Create(Notification.ServerWide, "Unable to start tcp listener", msg,
+                            AlertType.TcpListenerError, NotificationSeverity.Error, key: $"{ipAddress}:{port}", details: new ExceptionDetails(ex)));
```
Comment (ppekrol) [inline src/Raven.Server/RavenServer.cs]: should we add sth to the 'key' like a prefix? `tcp/listener/ip/port` maybe?

File: src/Raven.Server/Web/System/TestConnectionHandler.cs (hunk 14 above)
Comment (lastav5) [reply src/Raven.Server/Web/System/TestConnectionHandler.cs]: Fixed

File: src/Raven.Server/Documents/Handlers/Debugging/NodeDebugHandler.cs (hunk 11 above)
Comment (lastav5) [reply src/Raven.Server/Documents/Handlers/Debugging/NodeDebugHandler.cs]: The distinction isn't needed anymore. Fixed

File: src/Raven.Server/Documents/Handlers/Debugging/NodeDebugHandler.cs (hunk 10 above)
Comment (lastav5) [reply src/Raven.Server/Documents/Handlers/Debugging/NodeDebugHandler.cs:156]: Changed so PingResult doesn't allocate

File: src/Raven.Client/Util/TcpUtils.cs (hunk 9 above)
Comment (lastav5) [reply src/Raven.Client/Util/TcpUtils.cs]: Fixed

File: src/Raven.Client/Util/TcpUtils.cs (hunk 8 above)
Comment (lastav5) [reply src/Raven.Client/Util/TcpUtils.cs]: Done

File: src/Raven.Client/Util/TcpUtils.cs (hunk 7 above)
Comment (lastav5) [reply src/Raven.Client/Util/TcpUtils.cs]: Done

File: src/Raven.Client/ServerWide/Tcp/TcpConnectionHeader.cs (hunk 4 above)
Comment (lastav5) [reply src/Raven.Client/ServerWide/Tcp/TcpConnectionHeader.cs]: Deleted

File: src/Raven.Client/Util/TcpUtils.cs (hunk 8 above)
Comment (lastav5) [reply src/Raven.Client/Util/TcpUtils.cs]: Done

File: src/Raven.Server/Documents/Handlers/Debugging/NodeDebugHandler.cs (hunk 11 above)
Comment (lastav5) [reply src/Raven.Server/Documents/Handlers/Debugging/NodeDebugHandler.cs]: Done

File: src/Raven.Client/Util/TcpUtils.cs (hunk 2 above)
Comment (lastav5) [reply src/Raven.Client/Util/TcpUtils.cs]: Fixed

File: src/Raven.Client/Util/TcpUtils.cs (hunk 9 above)
Comment (lastav5) [reply src/Raven.Client/Util/TcpUtils.cs]: info.Url should be last

File: src/Raven.Server/ServerWide/ServerStore.cs (hunk 3 above)
Comment (lastav5) [reply src/Raven.Server/ServerWide/ServerStore.cs]: Done

File: src/Raven.Server/Documents/Handlers/Debugging/NodeDebugHandler.cs (hunk 12 above)
Comment (lastav5) [reply src/Raven.Server/Documents/Handlers/Debugging/NodeDebugHandler.cs]: Done

**Review by karmeli87 — APPROVED** [review]

File: src/Raven.Client/Util/TcpUtils.cs [hunk 16]
```diff
@@ -36,10 +40,14 @@ private static void SetTimeouts(TcpClient client, TimeSpan timeout)
             client.SendTimeout = (int)timeout.TotalMilliseconds;
             client.ReceiveTimeout = (int)timeout.TotalMilliseconds;
         }
-
-        internal static async Task<(TcpClient Client, string Url)> ConnectSocketAsync(
-            TcpConnectionInfo connection,
-            TimeSpan timeout,
+        internal static async Task<(TcpClient Client, Stream Stream, string Url, TcpConnectionHeaderMessage.SupportedFeatures SupportedFeatures)> ConnectSecuredTcpSocketAsReplication(
```
Comment (ayende) [inline src/Raven.Client/Util/TcpUtils.cs]: Better to create a return object, instead of that complex a tuple

File: src/Raven.Client/Util/TcpUtils.cs [hunk 17]
```diff
@@ -241,51 +256,99 @@ private static TcpClient NewTcpClient(TimeSpan? timeout, bool useIPv6)
...
+
+                    return (tcpClient, stream, url, supportedFeatures);
+                }
+                catch (Exception e)
+                {
+                    logs?.Add($"Failed to connect to url {url}: {e.Message}");
+                    if (i == infoUrls.Length - 1)
                     {
-                        // ignored
+                        throw;
```
Comment (ayende) [inline src/Raven.Client/Util/TcpUtils.cs]: Here, aggregate all the previous exceptions and throw an `AggregateException` if there is more than once.
We want to get the *full* errors for this scenario

File: src/Raven.Server/ServerWide/ClusterStateMachine.cs [hunk 18]
```diff
@@ -3480,6 +3468,23 @@ public override async Task<RachisConnection> ConnectToPeer(string url, string ta
...
+                Operation = TcpConnectionHeaderMessage.OperationTypes.Cluster,
+                Version = TcpConnectionHeaderMessage.ClusterTcpVersion,
+                ReadResponseAndGetVersionCallbackAsync = ClusterReadResponseAndGetVersion,
+                DestinationUrl = url,
+                DestinationNodeTag = tag,
+                SourceNodeTag = _parent.Tag,
+                DestinationServerId = info.ServerId
+            };
+
+            return await TcpNegotiation.NegotiateProtocolVersionAsync(ctx, stream, parameters);
```
Comment (ayende) [inline src/Raven.Server/ServerWide/ClusterStateMachine.cs:3513]: avoid the `await`, just return `Tsak`

File: src/Raven.Server/Web/System/TestConnectionHandler.cs [hunk 19]
```diff
@@ -41,62 +44,78 @@ await using (var writer = new AsyncBlittableJsonTextWriter(context, ResponseBody
...
+            
+            writer.WritePropertyName(nameof(TcpConnectionHeaderMessage.Operation));
+            writer.WriteString(operation.ToString());
+            writer.WriteComma();
+            writer.WritePropertyName(nameof(TcpConnectionHeaderMessage.OperationVersion));
+            writer.WriteInteger(TcpConnectionHeaderMessage.GetOperationTcpVersion(operation));
+            writer.WriteComma();
+            writer.WritePropertyName(nameof(TcpConnectionHeaderMessage.DatabaseName));
+            writer.WriteString(databaseName);
+            writer.WritePropertyName(nameof(TcpConnectionHeaderMessage.ServerId));
```
Comment (ayende) [inline src/Raven.Server/Web/System/TestConnectionHandler.cs:120]: Aren't you missing `WriteComma` here?

File: src/Raven.Server/ServerWide/ClusterStateMachine.cs (hunk 18 above)
Comment (lastav5) [reply src/Raven.Server/ServerWide/ClusterStateMachine.cs:3513]: But then it conflicts because the return type of NegotiateProtocolVersionAsync is ValueTask and not Task. And I'm not sure if I should make the return type of this whole callback function ValueTask

Comment (ppekrol) [reply src/Raven.Server/ServerWide/ClusterStateMachine.cs:3513]: You can, but then you need to be sure that this is awaited immediately in each place and not assigned to a variable

File: src/Raven.Server/Web/System/TestConnectionHandler.cs (hunk 19 above)
Comment (lastav5) [reply src/Raven.Server/Web/System/TestConnectionHandler.cs:120]: Yes. Fixed

File: src/Raven.Server/ServerWide/ClusterStateMachine.cs (hunk 18 above)
Comment (ayende) [reply src/Raven.Server/ServerWide/ClusterStateMachine.cs:3513]: Not a major problem, this should be rarely called.

File: src/Raven.Client/Util/TcpUtils.cs (hunk 17 above)
Comment (lastav5) [reply src/Raven.Client/Util/TcpUtils.cs]: The problem is that I have to keep the type of the error for outer functions that rely on it. For example ShouldTryToReconnect in subscriptions.

File: src/Raven.Client/Util/TcpUtils.cs (hunk 16 above)
Comment (lastav5) [reply src/Raven.Client/Util/TcpUtils.cs]: Done

File: src/Raven.Client/Documents/Subscriptions/SubscriptionWorker.cs [hunk 20]
```diff
@@ -278,6 +259,22 @@ await requestExecutor.ExecuteAsync(_redirectNode, null, context, command, should
             }
         }
 
+        private async Task<TcpConnectionHeaderMessage.SupportedFeatures> NegotiateProtocolVersionAsyncForSubscription(string chosenUrl, TcpConnectionInfo tcpInfo, Stream stream, JsonOperationContext context, List<string> _)
```
Comment (ppekrol) [inline src/Raven.Client/Documents/Subscriptions/SubscriptionWorker.cs]: `Async` in method names should be last. MS convention :)

File: src/Raven.Server/Web/System/TestConnectionHandler.cs [hunk 21]
```diff
@@ -41,62 +44,83 @@ await using (var writer = new AsyncBlittableJsonTextWriter(context, ResponseBody
...
 
                             case TcpConnectionStatus.TcpVersionMismatch:
                                 result.Success = false;
                                 result.Error = $"Connection to {url} failed because of mismatching tcp version: {headerResponse.Message}";
-                                await WriteOperationHeaderToRemote(writer, TcpConnectionHeaderMessage.OperationTypes.Drop, database);
-                                break;
+                                logs?.Add(result.Error);
+                                await WriteOperationHeaderToRemote(writer, TcpConnectionHeaderMessage.OperationTypes.Drop, database, info.ServerId);
+                                throw new AuthorizationException(result.Error);
+                            case TcpConnectionStatus.InvalidNetoworkTopology:
```
Comment (ppekrol) [inline src/Raven.Server/Web/System/TestConnectionHandler.cs]: typo

File: src/Raven.Server/RavenServer.cs [hunk 22]
```diff
@@ -1907,17 +1911,17 @@ private void ListenToNewTcpConnection(TcpListener listener)
                     if (_forTestingPurposes != null && _forTestingPurposes.ThrowExceptionInListenToNewTcpConnection)
                         throw new Exception("Simulated TCP failure.");
 
+                    var tcp = new TcpConnectionOptions
```
Comment (ppekrol) [inline src/Raven.Server/RavenServer.cs]: can this be moved back? there will be leak if sth will fail during AllocateOperationContext or GetMemoryBuffer below

File: src/Raven.Client/Exceptions/Security/InvalidNetworkTopologyException.cs [hunk 23]
```diff
@@ -0,0 +1,19 @@
+﻿using System;
+
+namespace Raven.Client.Exceptions.Security
+{
+    public class InvalidNetworkTopologyException : SecurityException
```
Comment (ppekrol) [inline src/Raven.Client/Exceptions/Security/InvalidNetworkTopologyException.cs]: is this a security exception?

File: test/RachisTests/SubscriptionFailoverWIthWaitingChains.cs [hunk 24]
```diff
@@ -222,7 +222,11 @@ public async Task SubscriptionShouldReconnectOnExceptionInTcpListener()
                 subsWorker.OnSubscriptionConnectionRetry += ex =>
                 {
                     Assert.NotNull(ex);
-                    Assert.True(ex.GetType() == typeof(IOException) || ex.GetType() == typeof(EndOfStreamException));
+                    
+                    if (ex is AggregateException ae)
```
Comment (ppekrol) [inline test/RachisTests/SubscriptionFailoverWIthWaitingChains.cs:226]: why those fixes were made? Asking because maybe we need to migrate them to 4.2 ?

File: src/Raven.Client/Exceptions/Security/InvalidNetworkTopologyException.cs (hunk 23 above)
Comment (lastav5) [reply src/Raven.Client/Exceptions/Security/InvalidNetworkTopologyException.cs]: @ayende Should this be a security exception or something else?

File: test/RachisTests/SubscriptionFailoverWIthWaitingChains.cs (hunk 24 above)
Comment (lastav5) [reply test/RachisTests/SubscriptionFailoverWIthWaitingChains.cs:226]: It's because ConnectSecuredSocket() (used to be connectAsync + WrapSsl + Negotiate) now returns an AggregateException instead of a single error. And because other places in this code rely on these errors I had to unwrap the AE and check each of them (otherwise tests fail)

File: src/Raven.Client/Exceptions/Security/InvalidNetworkTopologyException.cs (hunk 23 above)
Comment (ayende) [reply src/Raven.Client/Exceptions/Security/InvalidNetworkTopologyException.cs]: No, it should be inheriting from `Exception` directly, instead.

File: src/Raven.Client/Util/TcpUtils.cs (hunk 16 above)
Comment (ayende) [reply src/Raven.Client/Util/TcpUtils.cs]: That is still a tuple, no?

File: src/Raven.Server/Documents/Handlers/Debugging/NodeDebugHandler.cs (hunk 10 above)
Comment (ayende) [reply src/Raven.Server/Documents/Handlers/Debugging/NodeDebugHandler.cs:156]: ?

File: src/Raven.Server/Web/System/TestConnectionHandler.cs (hunk 19 above)
Comment (ayende) [reply src/Raven.Server/Web/System/TestConnectionHandler.cs:120]: ?

File: src/Raven.Client/Documents/Subscriptions/SubscriptionWorker.cs (hunk 20 above)
Comment (lastav5) [reply src/Raven.Client/Documents/Subscriptions/SubscriptionWorker.cs]: Done

File: src/Raven.Server/Web/System/TestConnectionHandler.cs (hunk 21 above)
Comment (lastav5) [reply src/Raven.Server/Web/System/TestConnectionHandler.cs]: Fixed

File: src/Raven.Client/Exceptions/Security/InvalidNetworkTopologyException.cs (hunk 23 above)
Comment (lastav5) [reply src/Raven.Client/Exceptions/Security/InvalidNetworkTopologyException.cs]: Fixed

File: src/Raven.Server/RavenServer.cs (hunk 22 above)
Comment (lastav5) [reply src/Raven.Server/RavenServer.cs]: Moved

File: src/Raven.Client/Documents/Subscriptions/SubscriptionWorker.cs (hunk 1 above)
Comment (ayende) [reply src/Raven.Client/Documents/Subscriptions/SubscriptionWorker.cs:279]: ?

File: src/Raven.Client/Util/TcpUtils.cs [hunk 25]
```diff
@@ -241,51 +257,117 @@ private static TcpClient NewTcpClient(TimeSpan? timeout, bool useIPv6)
             return tcpClient;
         }
 
-        internal static async Task<(TcpClient Client, string Url)> ConnectAsyncWithPriority(
-            TcpConnectionInfo info,
-            TimeSpan? tcpConnectionTimeout
+        internal class ConnectSecuredTcpSocketResult
```
Comment (ayende) [inline src/Raven.Client/Util/TcpUtils.cs]:
```suggestion
        internal struct ConnectSecuredTcpSocketResult
```
